controller/api/destination/external-workload/endpoints_reconciler.go

```diff
@@ -0,0 +1,624 @@
+}
+
+// newEndpointsReconciler takes an API client and returns a reconciler with
+// logging and a tracker set-up
+func newEndpointsReconciler(k8sAPI *k8s.API) *endpointsReconciler {
+	return &endpointsReconciler{
+		k8sAPI,
+		logging.WithFields(logging.Fields{
+			"component": "external-endpoints-reconciler",
+		}),
+		epsliceutil.NewEndpointSliceTracker(),
+	}
+
+}
+
+// === Reconciler ===
+
+// reconcile is the main entry-point for the reconciler's work.
+//
+// It accepts a slice of external workloads and their corresponding service.
+// Optionally, if the controller has previously created any slices for this
+// service, these will also be passed in. The reconciler will:
+//
+// * Determine what address families (AF) the service supports
+// * For each address family, it will determine which slices to process (an
+// EndpointSlice is specialised and supports only one AF type)
+func (r *endpointsReconciler) reconcile(svc *corev1.Service, ews []*ewv1alpha1.ExternalWorkload, es []*discoveryv1.EndpointSlice) error {
+	// segment slices by address type.
+	// 1. Find service's supported address types
+	// 2. Distribute slices based on what address family they are part of
+	// 3. If a slice has an address family that is no longer supported, it has
+	// to go.
+	addrTypes := getSupportedAddressTypes(svc)
+	toDelete := []*discoveryv1.EndpointSlice{}
+	ipv4Slices := []*discoveryv1.EndpointSlice{}
+
+	for _, slice := range es {
+		_, supported := addrTypes[slice.AddressType]
+		if !supported {
+			toDelete = append(toDelete, slice)
+			continue
+		}
+
+		// TODO: this could use a test
+		if slice.AddressType == discoveryv1.AddressTypeIPv4 {
+			ipv4Slices = append(ipv4Slices, slice)
+		}
+	}
+
+	// TODO (matei): we only process IPv4 for now. IPv6 support will go in here.
+	return r.reconcileIPv4Endpoints(svc, ews, ipv4Slices, toDelete)
+}
+
+// reconcileIPv4Endpoints operates on a set of external workloads, their
+// service, and any endpointslices that have been created by the controller. It
+// will compute the diff that needs to be written to the API Server.
+func (r *endpointsReconciler) reconcileIPv4Endpoints(svc *corev1.Service, extWorkloads []*ewv1alpha1.ExternalWorkload, epSlices []*discoveryv1.EndpointSlice, toDelete []*discoveryv1.EndpointSlice) error {
+
+	// We start the reconciliation by checking ownerRefs
+	//
+	// Note: in the upstream implementation, this codepath is different. Instead
+	// of appending to a slice of "endpoints to look at", they build a set of
+	// endpointslices keyed off the set of ports included in that endpointslice.
+	//
+	// Each slice may include a maximum of 100 ports. If a service includes more
+	// than 100 ports, then the slice won't be able to be filled. As a result,
+	// they segment slices by port sets.
+	// see: https://github.com/kubernetes/kubernetes/issues/99382
+	//
+	// We do not segment slices based on ports since we do not support a service
+	// with more than 100 ports.
```

Comment:
Do we need to segment sliced based on port for correctness if different EW have named ports which resolve to different ports?  e.g. consider a service which defines a port named `http`.  The service selects two EW, one with a container port named `http` with port value 8000 and the other EW has a container port named `http` with port value 8001.  Would these two EW need to go in different slices because their endpoints use different ports?

Reply:
Yup, good spot. Added in.